### PR TITLE
Add iterator for ivf_sq_8

### PR DESCRIPTION
**Pull request by cqy123456**
issue: https://github.com/zilliztech/knowhere/issues/425

### CODE REVIEW

File: thirdparty/faiss/faiss/IndexIVF.cpp [hunk 1]
```diff
@@ -933,6 +934,182 @@ void IndexIVF::range_search_preassigned(
...
+                    return (size_t)0;
+                }
+
+                // get scanner
+                IDSelector* sel = workspace->search_params
+                        ? workspace->search_params->sel
+                        : nullptr;
+                InvertedListScanner* scanner = get_InvertedListScanner(false, sel);
+                scanner->set_query(workspace->query_data);
+                scanner->set_list(list_no, coarse_list_centroid_dist);
```
Comment (cqy123456) [inline thirdparty/faiss/faiss/IndexIVF.cpp:975]: add a line :  add centroid id and distance for quantizer.

File: thirdparty/faiss/faiss/impl/ScalarQuantizerScanner.h [hunk 2]
```diff
@@ -83,6 +83,27 @@ struct IVFSQScannerIP : InvertedListScanner {
...
+            float* distances,
+            idx_t* labels,
+            size_t& counter_back) const override {
+        size_t nup = 0;
+        for (size_t j = 0; j < list_size; j++, codes += code_size) {
+            if (use_sel && !sel->is_member(use_sel == 1 ? ids[j] : j)) {
+                continue;
+            }
+            float accu = accu0 + dc.query_to_code(codes);
+            counter_back ++;
```
Comment (zhengbuqian) [inline thirdparty/faiss/faiss/impl/ScalarQuantizerScanner.h:101]: nit: space

File: thirdparty/faiss/faiss/impl/ScalarQuantizerScanner.h [hunk 3]
```diff
@@ -83,6 +83,27 @@ struct IVFSQScannerIP : InvertedListScanner {
...
+            heap_push<CMax<float, int64_t>>(counter_back, distances, labels, accu, ids[j]);
+            nup++;
+        }
+    }
```
Comment (zhengbuqian) [inline thirdparty/faiss/faiss/impl/ScalarQuantizerScanner.h:105]: missing return statement.

File: thirdparty/faiss/faiss/IndexIVF.cpp [hunk 4]
```diff
@@ -933,6 +934,182 @@ void IndexIVF::range_search_preassigned(
...
+                    size_t segment_offset =
+                            invlists->get_segment_offset(list_no, segment_idx);
+                    InvertedLists::ScopedCodes scodes(
+                            invlists, list_no, segment_offset);
+                    InvertedLists::ScopedCodeNorms scode_norms(
+                            invlists, list_no, segment_offset);
+                    InvertedLists::ScopedIds sids(
+                            invlists, list_no, segment_offset);
+
+                    size_t n_heap = scanner->scan_codes_and_push_back(
```
Comment (zhengbuqian) [inline thirdparty/faiss/faiss/IndexIVF.cpp:999]: temp variable n_heap is not used. discussed with @alwayslove2013 and we can remove it.

**Review by zhengbuqian — COMMENTED** [review]
thanks!

File: thirdparty/faiss/faiss/IndexIVF.cpp (hunk 4 above)
Comment (zhengbuqian) [reply thirdparty/faiss/faiss/IndexIVF.cpp]: sorry for being not precise in the previous comment: I meant to simply remove this local variable while `scan_codes_and_push_back` can still keep the return value for future use(could be used for metrics) and to have a aligned behavior with `scan_codes`.

File: thirdparty/faiss/faiss/IndexIVF.cpp [hunk 5]
```diff
@@ -933,6 +932,181 @@ void IndexIVF::range_search_preassigned(
...
+            auto list_size = invlists->list_size(list_no);
+            coarse_list_sizes[list_no] = list_size;
+            count += list_size;
+            if (list_size > workspace->max_coarse_list_size) {
+                workspace->max_coarse_list_size = list_size;
+            }
+        }
+
+        // compute backup_count_threshold - (nprobe / nlist) * count
+        size_t nprobe = workspace->search_params->nprobe
```
Comment (alexanderguzhva) [inline thirdparty/faiss/faiss/IndexIVF.cpp:1023]: what if `workspace->search_params` is `nullptr`?

File: thirdparty/faiss/faiss/IndexIVF.cpp [hunk 6]
```diff
@@ -933,6 +932,181 @@ void IndexIVF::range_search_preassigned(
...
 
+std::unique_ptr<IVFIteratorWorkspace> IndexIVF::getIteratorWorkspace(
+        const float* query_data,
+        const IVFSearchParameters* ivfsearchParams) const {
+    return std::make_unique<IVFIteratorWorkspace>(query_data, ivfsearchParams);
+}
+
+std::optional<std::pair<float, idx_t>> IndexIVF::getIteratorNext(
+        IVFIteratorWorkspace* workspace) const {
+    auto scan_one_list_then_add_to_backup =
```
Comment (alexanderguzhva) [inline thirdparty/faiss/faiss/IndexIVF.cpp:943]: please add a check if `workspace` is `nullptr`

File: thirdparty/faiss/faiss/IndexIVF.h [hunk 7]
```diff
@@ -89,6 +90,30 @@ struct InvertedListScanner;
 struct IndexIVFStats;
 struct CodePacker;
 
+struct IVFIteratorWorkspace {
```
Comment (alexanderguzhva) [inline thirdparty/faiss/faiss/IndexIVF.h:93]: Please add very detailed comments on what this structure does exactly and how it is used

File: thirdparty/faiss/faiss/IndexIVF.h [hunk 8]
```diff
@@ -313,6 +338,19 @@ struct IndexIVF : Index, IndexIVFInterface {
             RangeSearchResult* result,
             const SearchParameters* params = nullptr) const override;
 
+    std::unique_ptr<IVFIteratorWorkspace> getIteratorWorkspace(
```
Comment (alexanderguzhva) [inline thirdparty/faiss/faiss/IndexIVF.h:341]: Please add comments for this method about what it does

**Review by alexanderguzhva — CHANGES_REQUESTED** [review]

**Comment by zhengbuqian** [comment]
please hold until https://github.com/zilliztech/knowhere/pull/501 is merged. It slightly changed how we should implement iterator for indexes. It should make this PR much simpler.

**Pull request closed without merging**